src/main/java/gregtech/api/recipes/ingredients/GTRecipeInput.java

```diff
@@ -70,17 +72,33 @@ public int getAmount() {
      */
     public abstract GTRecipeInput copyWithAmount(int amount);
 
+    /**
+     * Returns either this instance with {@link GTRecipeInput#amount} field modified (for non-cached recipe inputs)
+     * or new copy with given amount (for cached recipe inputs).
+     */
+    public GTRecipeInput withAmount(int amount) {
+        if (getAmount() == amount) {
+            return this;
+        } else if (isCached()) {
+            return copyWithAmount(amount);
+        } else {
+            this.amount = amount;
+            return this;
+        }
+    }
+
     public GTRecipeInput setNonConsumable() {
-        GTRecipeInput copy = copy();
-        copy.isConsumable = false;
-        return getFromCache(copy);
+        if (!isConsumable) return this;
```

Comment:
Should a warning be logged if a not consumable ingredient is marked not consumable a second time?

Reply:
I don't think so.